ui/app/src/views/ViewDashboard.tsx

```diff
@@ -22,15 +22,45 @@ import {
   getDashboardExtendedDisplayName,
   DEFAULT_DASHBOARD_DURATION,
   DEFAULT_REFRESH_INTERVAL,
+  VariableResource,
 } from '@perses-dev/core';
-import { useCallback, useEffect, useRef, useState } from 'react';
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
 import { bundledPluginLoader } from '../model/bundled-plugins';
 import { useCreateDashboardMutation, useDashboard, useUpdateDashboardMutation } from '../model/dashboard-client';
 import DashboardBreadcrumbs from '../components/DashboardBreadcrumbs';
 import { useIsReadonly } from '../model/config-client';
 import { CreateAction } from '../model/action';
 import { CachedDatasourceAPI, HTTPDatasourceAPI } from '../model/datasource-api';
 import { useNavHistoryDispatch } from '../context/DashboardNavHistory';
+import { useVariableList } from '../model/project-client';
+
+/**
+ * The source type is limited in this part of the view. It can be:
+ * - 'project': variables defined for every dashboard of the current project.
+ * - 'global': variables defined for every dashboard of every project.
+ */
+type ExternalVariableSource = 'project' | 'global';
+
+/**
+ * Build the definition of the external variables from the variable resources collected from the API.
+ * @param source the source of external variables. Used in view only, to display and discriminate from other sources.
+ * @param variables variable resources collected from the API
+ */
+function buildExternalVariableDefinition(
```

Comment:
I like that this lives in the app since it's specific to the Perses app data model, but it'd be good to move out of ViewDashboard, ExternalVariableSource could go in `ui/app/src/model` and buildExternalVariableDefinition to `ui/app/src/utils`